// react-bootstrap-easy-dialog/dist/Dialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DialogProvider = DialogProvider;
exports.useDialog = useDialog;
exports.Dialog = Dialog;
exports.DialogUI = DialogUI;
exports.DialogConsumer = exports.DialogContext = void 0;

var _react = _interopRequireWildcard(require("react"));

var _Modal = _interopRequireDefault(require("react-bootstrap/Modal"));

var _Button = _interopRequireDefault(require("react-bootstrap/Button"));

var _Form = _interopRequireDefault(require("react-bootstrap/Form"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var DialogContext = (0, _react.createContext)(null);
exports.DialogContext = DialogContext;

function DialogProvider(_ref) {
  var children = _ref.children,
      providerOptions = _objectWithoutProperties(_ref, ["children"]);

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      show = _useState2[0],
      setShow = _useState2[1];

  var _useState3 = (0, _react.useState)({}),
      _useState4 = _slicedToArray(_useState3, 2),
      showOptions = _useState4[0],
      setShowOptions = _useState4[1];

  var _useState5 = (0, _react.useState)(true),
      _useState6 = _slicedToArray(_useState5, 2),
      hidden = _useState6[0],
      setHidden = _useState6[1];

  var hiddenRef = (0, _react.useRef)(hidden);
  var previousHiddenRef = (0, _react.useRef)(hidden);
  var resolveDoneRef = (0, _react.useRef)(null);
  var resolveHiddenRef = (0, _react.useRef)(null);
  var doneResultRef = (0, _react.useRef)();

  var mergedOptions = _objectSpread({}, providerOptions, {}, showOptions, {
    show: show
  });

  function resolveDone(result) {
    if (resolveDoneRef.current) {
      resolveDoneRef.current(result);
      resolveDoneRef.current = null;
      doneResultRef.current = result;
    }
  }

  function resolveHidden() {
    if (resolveHiddenRef.current) {
      resolveHiddenRef.current(doneResultRef.current);
      resolveHiddenRef.current = null;
      doneResultRef.current = undefined;
    }
  }

  (0, _react.useEffect)(function () {
    hiddenRef.current = hidden;
    var previousHidden = previousHiddenRef.current;
    previousHiddenRef.current = hidden;

    if (previousHidden === false && hidden === true) {
      resolveHidden();
    }
  }, [hidden]);
  var dialog = (0, _react.useMemo)(function () {
    function buildMethod(methodOptions, failValue, okValue) {
      return function (text, userOptions) {
        if (!hiddenRef.current) return failValue;
        setShow(true);
        setHidden(false);
        setShowOptions(_objectSpread({}, methodOptions, {}, userOptions, {
          text: text,
          onCancel: function onCancel() {
            setShow(false);
            resolveDone(failValue);
          },
          onConfirm: function onConfirm(result) {
            setShow(false);
            resolveDone(okValue === undefined ? result : okValue);
          },
          onExited: function onExited() {
            setShowOptions({});
            setHidden(true);
          }
        }));
        var donePromise = new Promise(function (resolve) {
          resolveDoneRef.current = resolve;
        });
        donePromise.done = donePromise;
        donePromise.hidden = new Promise(function (resolve) {
          resolveHiddenRef.current = resolve;
        });
        return donePromise;
      };
    }

    return {
      alert: buildMethod({
        input: false,
        cancelButton: false,
        confirmButton: true
      }, false, true),
      confirm: buildMethod({
        input: false,
        cancelButton: true,
        confirmButton: true
      }, false, true),
      prompt: buildMethod({
        input: true,
        cancelButton: true,
        confirmButton: true
      }, null, undefined)
    };
  }, []);
  return _react["default"].createElement(DialogContext.Provider, {
    value: dialog
  }, children, _react["default"].createElement(DialogUI, mergedOptions));
}

var DialogConsumer = DialogContext.Consumer;
exports.DialogConsumer = DialogConsumer;

function useDialog() {
  return (0, _react.useContext)(DialogContext);
}

function Dialog(_ref2) {
  var children = _ref2.children,
      options = _objectWithoutProperties(_ref2, ["children"]);

  return _react["default"].createElement(DialogProvider, options, _react["default"].createElement(DialogConsumer, null, children));
}

function DialogUI(_ref3) {
  var show = _ref3.show,
      title = _ref3.title,
      text = _ref3.text,
      input = _ref3.input,
      inputProps = _ref3.inputProps,
      cancelButton = _ref3.cancelButton,
      cancelButtonProps = _ref3.cancelButtonProps,
      confirmButton = _ref3.confirmButton,
      confirmButtonProps = _ref3.confirmButtonProps,
      onConfirm = _ref3.onConfirm,
      onCancel = _ref3.onCancel,
      onExited = _ref3.onExited,
      _ref3$autoFocus = _ref3.autoFocus,
      autoFocus = _ref3$autoFocus === void 0 ? "select" : _ref3$autoFocus,
      _ref3$stubborn = _ref3.stubborn,
      stubborn = _ref3$stubborn === void 0 ? false : _ref3$stubborn;

  var _ref4 = inputProps || {},
      _ref4$defaultValue = _ref4.defaultValue,
      defaultValue = _ref4$defaultValue === void 0 ? "" : _ref4$defaultValue,
      _ref4$refKey = _ref4.refKey,
      refKey = _ref4$refKey === void 0 ? "ref" : _ref4$refKey,
      otherInputProps = _objectWithoutProperties(_ref4, ["defaultValue", "refKey"]);

  var _useState7 = (0, _react.useState)(""),
      _useState8 = _slicedToArray(_useState7, 2),
      inputValue = _useState8[0],
      setInputValue = _useState8[1];

  var inputRef = (0, _react.useRef)();
  var submitButtonRef = (0, _react.useRef)(); // only set input value to default value when it shows up

  (0, _react.useEffect)(function () {
    if (show) {
      setInputValue(defaultValue);
    }
  }, [show, defaultValue]); // set autoFocus properly

  (0, _react.useEffect)(function () {
    if (show && input && autoFocus) {
      inputRef.current.focus();

      if (autoFocus === "select") {
        setImmediate(function () {
          inputRef.current && inputRef.current.select();
        });
      }
    }
  }, [show, input, autoFocus]);
  var showHeader = !!title;
  var showBody = !!text || !!input;

  function handleSubmit(e) {
    e.preventDefault();
    onConfirm(inputValue);
  }

  function handleConfirm() {
    if (submitButtonRef.current) {
      submitButtonRef.current.click();
    } else {
      onConfirm();
    }
  }

  function handleHide() {
    if (!stubborn) {
      handleCancel();
    }
  }

  function handleCancel() {
    onCancel();
  }

  function handleInputValueChange(e) {
    setInputValue(e.target.value);
  }

  function handleExited() {
    // clear the input value on finish of the animation
    setInputValue("");
    onExited();
  }

  return _react["default"].createElement(_Modal["default"], {
    show: show,
    onHide: handleHide,
    onExited: handleExited
  }, showHeader && _react["default"].createElement(_Modal["default"].Header, null, _react["default"].createElement(_Modal["default"].Title, null, title)), showBody && _react["default"].createElement(_Modal["default"].Body, null, !!text && _react["default"].createElement("p", {
    style: {
      marginBottom: input ? undefined : 0
    }
  }, text), input && _react["default"].createElement(_Form["default"], {
    onSubmit: handleSubmit
  }, _react["default"].createElement(_Form["default"].Control, _extends({}, otherInputProps, {
    value: inputValue,
    onChange: handleInputValueChange
  }, _defineProperty({}, refKey, inputRef))), _react["default"].createElement(_Button["default"], {
    type: "submit",
    hidden: true,
    ref: submitButtonRef
  }, "Submit"))), _react["default"].createElement(_Modal["default"].Footer, null, cancelButton && _react["default"].createElement(_Button["default"], _extends({
    variant: "secondary",
    children: "Cancel"
  }, cancelButtonProps, {
    onClick: handleCancel
  })), confirmButton && _react["default"].createElement(_Button["default"], _extends({
    variant: "primary",
    children: "Confirm"
  }, confirmButtonProps, {
    onClick: handleConfirm
  }))));
}
